mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala

```diff
@@ -261,31 +264,50 @@ class BisectingKMeans @Since("2.0.0") (
   @Since("2.4.0")
   def setDistanceMeasure(value: String): this.type = set(distanceMeasure, value)
 
+  /**
+   * Sets the value of param [[weightCol]].
+   * If this is not set or empty, we treat all instance weights as 1.0.
+   * Default is not set, so all instances have weight one.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): BisectingKMeansModel = instrumented { instr =>
     transformSchema(dataset.schema, logging = true)
 
     val handlePersistence = dataset.storageLevel == StorageLevel.NONE
-    val rdd = DatasetUtils.columnToOldVector(dataset, getFeaturesCol)
+    val w = if (isDefined(weightCol) && $(weightCol).nonEmpty) {
+      col($(weightCol)).cast(DoubleType)
+    } else {
+      lit(1.0)
+    }
+
+    val instances: RDD[(OldVector, Double)] = dataset
+      .select(DatasetUtils.columnToVector(dataset, getFeaturesCol), w).rdd.map {
+      case Row(point: Vector, weight: Double) => (OldVectors.fromML(point), weight)
+    }
     if (handlePersistence) {
```

Comment:
Maybe we no longer need to (or should not) handlePersistence in the .ml sides? for KMeans & BisectingKMeans?  @srowen 

No matter whether input dataset is cached, we always cache a training rdd (`Vector`s in .ml, or `VectorWithNorm`s in .mlllib). So we should cache the one with norms for better performace.

Reply:
The persistence strategy is inconsistent. I'd prefer to standardize it more than change it though. We want to avoid silently taking a big performance hit by not caching. Either that means warning the user or caching internal RDDs for the user.

Reply:
I guess we should suggest end uers NOT to cache input DFs/RDDs, and the impls should always cache necessary internal DFs/RDDs.
Some time user caching DFs outside of impl will not help performance too much, for example in decision tree, vectors will be convert to treepoints, and then the cached vectors are not used any more.